metricbeat/module/haproxy/haproxy.go

```diff
@@ -0,0 +1,243 @@
+	SslCacheMisses             string `mapstructure:"SslCacheMisses"`
+	CompressBpsIn              string `mapstructure:"CompressBpsIn"`
+	CompressBpsOut             string `mapstructure:"CompressBpsOut"`
+	CompressBpsRateLim         string `mapstructure:"CompressBpsRateLim"`
+	ZlibMemUsage               string `mapstructure:"ZlibMemUsage"`
+	MaxZlibMemUsage            string `mapstructure:"MaxZlibMemUsage"`
+	Tasks                      string `mapstructure:"Tasks"`
+	RunQueue                   string `mapstructure:"Run_queue"`
+	IdlePct                    string `mapstructure:"Idle_pct"`
+	Node                       string `mapstructure:"Node"`
+	Description                string `mapstructure:"Description"`
+}
+
+// Client is an instance of the HAProxy client
+type Client struct {
+	Address     string
+	ProtoScheme string
+}
+
+// NewHaproxyClient returns a new instance of HaproxyClient
+func NewHaproxyClient(address string) (*Client, error) {
+	parts := strings.Split(address, "://")
+	if len(parts) != 2 {
+		return nil, errors.New("Must have protocol scheme and address!")
+	}
+
+	if parts[0] != "tcp" && parts[0] != "unix" {
+		return nil, errors.New("Invalid Protocol Scheme!")
+	}
+
+	return &Client{
+		Address:     parts[1],
+		ProtoScheme: parts[0],
+	}, nil
+}
+
+// Run sends a designated command to the haproxy stats socket
+func (c *Client) run(cmd string) (*bytes.Buffer, error) {
+	var conn net.Conn
+	response := bytes.NewBuffer(nil)
+
+	conn, err := net.Dial(c.ProtoScheme, c.Address)
+	if err != nil {
+		return response, err
+	}
+
+	defer conn.Close()
+
+	_, err = conn.Write([]byte(cmd + "\n"))
+	if err != nil {
+		return response, err
+	}
+
+	_, err = io.Copy(response, conn)
+	if err != nil {
+		return response, err
+	}
+
+	if strings.HasPrefix(response.String(), "Unknown command") {
+		return response, fmt.Errorf("Unknown command: %s", cmd)
+	}
+
+	return response, nil
+}
+
+// GetStat returns the result from the 'show stat' command
+func (c *Client) GetStat() ([]*Stat, error) {
+
+	runResult, err := c.run("show stat")
+	if err != nil {
+		return nil, err
+	}
+
+	var statRes []*Stat
+	csvReader := csv.NewReader(runResult)
+	csvReader.TrailingComma = true
+
+	err = gocsv.UnmarshalCSV(csvReader, &statRes)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing CSV: %s", err)
+	}
+
+	return statRes, nil
+
+}
+
+// GetInfo returns the result from the 'show stat' command
+func (c *Client) GetInfo() (*Info, error) {
+
+	res, err := c.run("show info")
+	if err != nil {
+		return nil, err
+	}
+
+	if b, err := ioutil.ReadAll(res); err == nil {
+
+		resultMap := map[string]interface{}{}
+
+		for _, ln := range strings.Split(string(b), "\n") {
+
+			ln := strings.TrimSpace(ln)
+			if ln == "" {
+				continue
+			}
+
+			parts := strings.Split(ln, ":")
+			if len(parts) != 2 {
+				continue
+			}
+
+			resultMap[parts[0]] = strings.TrimSpace(parts[1])
+		}
+
+		var result *Info
+		err := mapstructure.Decode(resultMap, &result)
+		if err != nil {
```

Comment:
Consider writing this which is slightly more concise. (not a mandatory change, it's just how I'd of written it)

```
if err := mapstructure.Decode(resultMap, &result); err != nil {
    return nil, err
}
```